### ISSUE TITLE
Freezing layers with `eqx.tree_map` (not parameters in a single layer) 

### ISSUE THREAD

**Issue by homerjed** [patrick-kidger/equinox#797]
Hello all,

Running around in circles a bit here. I'm trying to implement the 'Stochastic Depth' layer in the [Compact Convolutional Transformers](https://keras.io/examples/vision/cct/#stochastic-depth-for-regularization) paper.

This is basically dropout applied to a sequence of layers as opposed to elements of a weight matrix.

To do this, I thought I could tree-map-mask the pytree list of layers, somehow indexing a random percentage of them to target (akin to dropout).

I'll save pasting a lot of useless workings, but I thought something like this should work

```python

key = jr.key(0)

# Layers to drop (should end up with all the layers after, but some with all arrays having their gradients stopped)
layers = tuple([eqx.nn.Linear(1, 1, key=key) for _ in range(4)])

rate = 0.5 # Drop half of the layers
ix_of_layers_to_drop = jr.choice(key, jnp.arange(len(layers)), (int(rate * len(layers)),))

layers_ = eqx.tree_at(
   # Return nodes of layers (that are linear layers themselves) that should be replaced in layers with 'frozen' equivalents
    lambda layers: [layers[i] for i in ix_of_layers_to_drop], 
    # Only stop gradients of arrays
    eqx.filter(layers, eqx.is_array),
    # Replace leaves with 'frozen' leaves
    replace_fn=lambda x: jax.lax.stop_gradient(x)
)
assert len(layers_) == len(layers) # Not really testing much...
```

does this make sense? 

One hiccup may be that `jax.lax.stop_gradient` works on PyTrees, meaning that the replace function directly replaces entire linear layers in this case (the array-leaves of them at least!). I've applied this function to the array-only PyTrees of the linear layers which should be correct.

Cheers
Jed

**Comment by patrick-kidger** [comment]
It looks lke you have a tracer (`ix_of_layers_to_drop`) that you're using to try and index into a list (`layers[i]`). You've not mentioend what your error message is but I suspect this is what you're bumping into?

(Side note, I don't see the connection between dropout and freezing layers -- from your link it looks more like certain values are randomly set to zero.)

**Comment by homerjed** [comment]
Sorry totally missed how this works, fixed it with the true method of the module. Cheers!
